docs/source/extending/overloading-guide.rst

```diff
@@ -0,0 +1,251 @@
+  for the user should be issued.
+
+The resulting implementation could look as follows:
+
+.. literalinclude:: myjitmodule2.py
+
+As you can see, the typing checking code has been increased significantly to
+match the new requirements. Also, multiple implementations---one for integers
+and one for floating-point---are provided. We check inside the typing scope
+which implementation should be used and also raise any custom error messages
+required. Importantly, the check for ``nan`` values is only present in the
+floating point implementation as this additional check creates a runtime
+overhead. This can easily be observed during profiling:
+
+.. code:: pycon
+
+    In [1]: a = np.arange(100, dtype=np.float64)
+
+    In [2]: %timeit myalgorithm(a, 1.0)
+    473 ns ± 11.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
+
+    In [3]: a = np.arange(100)
+
+    In [4]: %timeit myalgorithm(a, 1)
+    237 ns ± 4.59 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
+
+Writing Tests
+=============
+
+The following is a sufficient piece of test code for the overloaded
+``set_to_x`` implementation. As you can see, only a small part of the test-code
+is about testing if the function works correctly. Most of the test code in this
+example checks that all error cases are handled and that all raised exceptions
+are of the correct type and have the correct error message. When implementing
+tests cases for Numba, you should always use ``numba.tests.support.TestCase``.
+
+.. literalinclude:: test_myjitmodule2.py
+
+While is is a pretty descent test, it wouldn't be accepted into the Numba.
+There are a few more test-cases that should be implemented:
+
+* Empty array
+* Single value array
+* Multidimensional arrays
+* Tests for ``int32`` and ``float32`` types
+
+You can download the example code from above including tests here:
+
+* :download:`mymodule.py </extending/mymodule.py>`
+* :download:`myjitmodule.py </extending/myjitmodule.py>`
+* :download:`myjitmodule2.py </extending/myjitmodule2.py>`
+* :download:`test_myjitmodule2.py </extending/test_myjitmodule2.py>`
+
+Implementing ``overload`` for Numpy functions
+=============================================
+
+Numba has strong support for NumPy because it provides ``jit`` compatible
+re-implementations of NumPy functions. In such cases ``overload`` is a very
+convenient option, however there are a few additional things to watch out for.
+
+* The  Numba implementation should match the NumPy implementation as closely as
+  feasible with respect to accepted types, arguments, raised exceptions and
+  runtime (Big-O / Landau order).
+
+* If you are implementing a new function, you should always update the
+  `documentation
+  <http://numba.pydata.org/numba-doc/latest/reference/numpysupported.html>`_.
+  The sources can be found in `docs/source/reference/numpysupported.rst``. Be
+  sure to mention any limitations that your implementation has, e.g. no support
+  for the ``axis`` keyword.
+
```

Comment:
Maybe add a bullet point along the lines that numpy accepts many types of input, generally (scalar, list, tuple, array, generator, tuple of lists etc) which you need to bear in mind in testing and implementation.

Further: a numpy function may return a scalar, array or a data structure which matches one of its inputs - so need to watch out for type unification problems and dispatch to appropriate implementations.

CI tests a wide variety of numpy versions - you'll sometimes be alerted to a change in behaviour back in some previous numpy version.  If you can find supporting evidence in the numpy change log / repo, then you'll need to decide whether to branch logic and attempt to replicate the logic across versions, or use a version gate (with associated wording in the docs) to advertise that numba replicates numpy from v1.11 onwards, for example. 

When writing tests, it's useful to include handling of non-finite values; arrays with different shapes and layouts; complex inputs; scalar inputs; inputs with types for which support is not documented (e.g. a function which the numpy docs say requires a float or int input might also 'work' if given a bool or complex input).

Reply:
These are all very good points. I took your suggestions and tweaked them slightly and included them. Do you perhaps have a good example for the second point?

Reply:
Hi, with regards to the second point, maybe something like `np.corrcoef` - which either returns a scalar or an array depending on the inputs - is a reasonable example.  

I guess it's not much different to your example of dispatching based on input dtype; it's just a relatively common motivator for writing multiple implementations and avoiding type unification problems in so doing.

Reply:
That's useful, thanks!